### PR TITLE
Implement step 5 of select_image_source

### LINKED ISSUE #11416: Support responsive images
This is a large piece of work that probably depends on making our image loading code better reflect the actual specification model. For more context on what responsive images are, see [the specification introduction](https://html.spec.whatwg.org/multipage/embedded-content.html#introduction-3).

Components:
- [x] [`<picture>`](https://html.spec.whatwg.org/multipage/embedded-content.html#the-picture-element)
- [x] [`<source>`](https://html.spec.whatwg.org/multipage/embedded-content.html#the-source-element)
- [x] [sizes attribute](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-sizes) (see #10989 for prior work)
- [x] [srcset attribute](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-srcset) (see #10829 for prior work)
- [x] [relevant mutations](https://html.spec.whatwg.org/multipage/embedded-content.html#relevant-mutations) processing
- [x] [update the source set](https://html.spec.whatwg.org/multipage/embedded-content.html#update-the-source-set) as part of the image selection process
- [x] [environment changes](https://html.spec.whatwg.org/multipage/embedded-content.html#img-environment-changes) processing

### PR DESCRIPTION
**Pull request by nupurbaghel**
This step is responsible for selecting an image source based on user agent's environment( closest value which is greater than user's pixel density) 
In case of no device, maximum density source is selected.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #11416 

- [x] These changes do not require tests because tests are already present

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21215)

### CODE REVIEW

**Comment by jdm** [comment]
That's #20234.

**Comment by nupurbaghel** [comment]
Alright, this issue seems to be a problem to everyone. How do you solve it 😅 ?

File: components/script/dom/htmlimageelement.rs [hunk 1]
```diff
@@ -603,15 +603,38 @@ impl HTMLImageElement {
...
         for outer_index in 0..len {
             if !repeat_indices.contains(&outer_index) {
                 img_sources.push(&source_set.image_sources[outer_index]);
+                // set max density and index
+                let den = img_sources[current_index].descriptor.den.unwrap();
+                if max_den < den {
+                    max_den = den;
+                    max_index = current_index;
+                }
+                current_index = current_index + 1;
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: We can write this more idiomatically ask:
```rust
let mut max = (0f64, 0);
for image_source in &source_set.image_sources {
    if repeat_indices.contains(&index) {
        continue;
    }
    let den = image_source.descriptor.den.unwrap();
    if max.0 < den {
        max = (den, img_sources.len());
    }
    img_sources.push(image_source);
}
```

File: components/script/dom/htmlimageelement.rs [hunk 2]
```diff
@@ -603,15 +603,38 @@ impl HTMLImageElement {
...
 
-        // TODO Step 5 - select source based on pixel density
-        let selected_source = img_sources.remove(0).clone();
+        // Step 5
+        let mut closest_den = max_den;
+        let mut result_index = max_index;
+        let device = document_from_node(self).device();
+        if device.is_some() {
+            let device_den: f64 = device.unwrap().device_pixel_ratio().get().into();
+            for outer_index in 0..img_sources.len() {
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: We can write this more idiomatically as:
```rust
if let Some(device) = device {
    let device_den = device.device_pixel_ratio().get() as f64;
    let mut best_candidate = max;
    for (index, image_source) in img_sources.iter().enumerate() {
        let current_den = image_source.descriptor.den.unwrap();
        if current_den < best_candidate.0 && current_den >= device_den {
            best_candidate = (current_den, index);
        }
    }
}
```

**Review by jdm — CHANGES_REQUESTED** [review]
My only suggestion here is to use iterators for these loops instead of explicit array indexes when possible.

File: components/script/dom/htmlimageelement.rs (hunk 1 above)
Comment (nupurbaghel) [reply components/script/dom/htmlimageelement.rs]: Do we have to generate `index` here too by enumerating like the next suggestion ?

Comment (jdm) [reply components/script/dom/htmlimageelement.rs]: Oh, right. Yes, I missed that one.

**Comment by nupurbaghel** [comment]
Made the suggested changes 👍

**Review by jdm — APPROVED** [review]

**Pull request merged**